File: fsharp/AdventOfCode/Puzzles/Day07.fs
namespace AdventureOfCode.Puzzles

open System
open System.IO

module Day07 =
    
    let triangularNumber n = (n*(n+1))/2

    let parseInput filePath =
        let numbers =
            File.ReadAllText filePath
            |> List.ofSeq
            |> List.map (Operators.int)

        numbers

    let solvePart1 filePath =
        let numbers = parseInput filePath
        let min = Seq.min numbers
        let max = Seq.max numbers

        let minimumFuelConsumption = 
            seq { min .. max }
            |> Seq.map
                (fun step ->
                    numbers
                    |> Seq.map (fun number -> Math.Abs(number - step))
                    |> Seq.sum)
            |> Seq.min
        minimumFuelConsumption 

    let solvePart2 filePath =
        let numbers = parseInput filePath
        Seq.min numbers

File: csharp/AdventOfCode.Core/Puzzles/Day07.cs
namespace AdventOfCode.Core.Puzzles;

public class Day07
{
    public object SolvePart1(string inputFile)
    {
        var input = InputHelper.ReadLinesFromFile(inputFile);
        var positions = input[0].Split(',').Select(int.Parse).ToArray();

        return Enumerable
            .Range(positions.Min(), positions.Max())
            .Select(number => positions.Sum(item => Math.Abs(number - item)))
            .Prepend(int.MaxValue)
            .Min();
    }

    public object SolvePart2(string inputFile)
    {
        var input = InputHelper.ReadLinesFromFile(inputFile);
        var positions = input[0].Split(',').Select(int.Parse).ToArray();

        return Enumerable
            .Range(positions.Min(), positions.Max())
            .Select(number => positions.Select(item => Math.Abs(number - item))
                .Select(distance => distance * (distance + 1) / 2)
                .Sum())
            .Prepend(int.MaxValue)
            .Min();
    }

}